### DISCUSSION TITLE
Idle pin state in PWM

### DISCUSSION THREAD

**Discussion by DonatelloX** [stm32duino/Arduino_Core_STM32#2447] [Q&A] [answered]
Arduino_Core_STM32 2.8.0 + Arduino IDE 2.3.2
STM32F303CC in custom board

I use this part of code to initialize a PWM:

```
  HardwareTimer *timerTreatment;
  [...]
  pinMode(pwmPinTreatment, OUTPUT);
  TIM_TypeDef *Instance = (TIM_TypeDef *)pinmap_peripheral(digitalPinToPinName(pwmPinTreatment), PinMap_PWM);
  uint32_t channel = STM_PIN_CHANNEL(pinmap_function(digitalPinToPinName(pwmPinTreatment), PinMap_PWM));
  timerTreatment = new HardwareTimer(Instance);
  timerTreatment->setPWM(channel, pwmPinTreatment, baseFrequencyTreatment, 53);
  timerTreatment->pause();
  [...]
```

PWM works well. But when I use it in my various parts of the code, doing:
```
[...]
timerTreatment->resume();
[...]
timerTreatment->pause();
[...]
```
after pause() the PWM signal in the idle state is sometimes at a logical high level (wrong) and sometimes correctly at a low (correct) level. It should always be low when pause().
I can't use it properly.

**Comment by fpistm** [comment]
Hi @DonatelloX 
There is no issue.
When you pause the HT, you simply pause the counter. So the pin can have High or low state depending of the counter value.
When you resume it continues from the last count value.

If you want it at low state then try to use the `setCount()` api to set it to 0.

**Comment by DonatelloX** [comment, marked as answer]
Thank you.
I use:
`timerTreatment->setCount(timerTreatment->getOverflow()-1);`
after:
`timerTreatment->pause();`
and I think it works now.

I have changed your pause() routine in HardwareTimer.cpp:
```
void HardwareTimer::pause(bool setLow = false)
{
 // [...]

  if (setLow) {
    uint32_t maxCount = getOverflow();
    setCount(maxCount - 1);
  }

}
```

**Reply by fpistm** [reply]
Fine, but pay attention next time you will update the core your change will be overwritten

**Reply by DonatelloX** [reply]
@fpistm 
I understand that it is a timer and therefore pause() stops the timer, but with PWM the signal in pause() must return to its idle state.
My advice: it would be much better if, when using setPWM, it is specified that a PWM is desired and therefore, if the pause() function is used, the library sets the correct idle state.

**Reply by fpistm** [reply]
Well from your point of view is the desired way but not for other. Some simply wants restarts where it stop, some other would high level.... Moreover in that case, the pause is requested so application have to know what it does.
